### PR TITLE
Drug labels in pml

### PR DESCRIPTION
**Pull request by c-brenn**
new drug format as per customer's requests.

### CODE REVIEW

**Comment by c-brenn** [comment]
drugs can be specified in PML like so
```
requires { drug { "paracetamol" } }
```

* `/api/pml` returns the list of english drugs in a pml file
* `/api/uris` takes a list of english drugs and uses the new asclpius endpoints to find their uris
* `api/ddis` takes a list of uris and returns their ddis

The front-end code works but is nice and ugly. The `content-type` has to be explicitly set to `application/json` or else phoenix thinks it's `text/plain`

File: panacea/web/static/js/app.js [hunk 1]
```diff
@@ -12,17 +12,48 @@ async function submitFile() {
...
+        method: 'POST',
+        body: JSON.stringify({labels: drugs}),
+        credentials: 'same-origin',
+        headers: new Headers({authorization: apiAccessToken, "Content-Type": "application/json"})
+      });
+
+      if (uris_response.ok) {
+        const data = await uris_response.json();
+        const uris = Object.keys(data.uris.found);
+        console.log(uris);
```
Comment (c-brenn) [inline panacea/web/static/js/app.js]: will remove this

File: asclepius/README.md [hunk 2]
```diff
@@ -82,28 +82,57 @@ Endpoints
...
+| Methods      | `POST`                                                                                                           |
+| Request Body | An object containing a list of drug labels, named `labels`                                                       |
+| Returns      | See below                                     |
+
+#### Example
+
+##### Request Body
+
+```json
+{"labels": [ "paracetamol", "flat seven up", "cocaine"]}
```
Comment (houli) [inline asclepius/README.md:121]: Does this need to be labels or can we just send the list with no object around it?

File: asclepius/README.md [hunk 3]
```diff
@@ -82,28 +82,57 @@ Endpoints
 | Returns      | A list of DDI objects; its label, its URI, and the identifiers of the two drugs involved                                         |
 
 #### Example
-
 ##### Request Body
-
 ```json
-{"drugs": ["chebi:421707", "chebi:465284", "dinto:DB00503", "chebi:9342"]}
+{
+  "drugs": [
```
Comment (houli) [inline asclepius/README.md:88]: Does this need to be drugs or can we just send the list with no object around it? Realise this is an old one but sure

File: asclepius/README.md [hunk 4]
```diff
@@ -82,28 +82,57 @@ Endpoints
...
+
+```json
+{"labels": [ "paracetamol", "flat seven up", "cocaine"]}
+```
+
+##### Response Body
+
+```json
+{
+  "found": {
```
Comment (houli) [inline asclepius/README.md]: Maybe we want this instead
```json
{
  "found": [
    {
      "uri": "http://purl.obolibrary.org/obo/CHEBI_27958",
      "label": "cocaine"
    },
    {
      "uri": "http://purl.obolibrary.org/obo/CHEBI_46195",
      "label": "paracetamol"
    }
  ],
  "not_found": [
    "flat seven up"
  ]
}
```

File: asclepius/asclepius/dinto.py [hunk 5]
```diff
@@ -57,14 +57,25 @@ def sparqled(*args, **kwargs):
     return sparqled
 
 
+@lru_cache()
 @sparql
-def all_drugs():
+def drugs(labels=None):
+
+    if labels is not None:
+        if not isinstance(labels, frozenset):
```
Comment (houli) [inline asclepius/asclepius/dinto.py]: Not sure we need to catch this, we should accept any hashable, if it's not hashable then the lru_cache will throw anyway

File: asclepius/asclepius/dinto.py [hunk 6]
```diff
@@ -81,22 +92,16 @@ def all_ddis():
...
-def _valid_drug(drug_identifier):
-    return DRUG_PATTERN.match(drug_identifier) is not None
-
-
 @sparql
 def ddi_from_drugs(drugs):
     if not isinstance(drugs, frozenset):
         raise ValueError("for cachability, `drugs` must be given as a frozenset")
-
+    print(drugs)
```
Comment (houli) [inline asclepius/asclepius/dinto.py]: Presume this isn't needed

File: panacea/web/static/js/app.js [hunk 7]
```diff
@@ -12,17 +12,47 @@ async function submitFile() {
   hideFileForm();
   hideResults();
   try {
-    const response = await fetch(this.action, {
+    const drugs_response = await fetch(this.action, {
```
Comment (houli) [inline panacea/web/static/js/app.js]: A style thing for the JS bits camelCase it

File: panacea/web/static/js/app.js [hunk 8]
```diff
@@ -12,17 +12,47 @@ async function submitFile() {
...
       method: 'POST',
       body: new FormData(this),
       credentials: 'same-origin',
       headers: new Headers({authorization: apiAccessToken})
```
Comment (houli) [inline panacea/web/static/js/app.js]: We'll probably want `'Accept': 'application/json, text/plain, */*'` headers on all of these requests

**Review by houli — CHANGES_REQUESTED** [review]

File: asclepius/README.md (hunk 4 above)
Comment (c-brenn) [reply asclepius/README.md]: our reasoning for the current format is that when DDIs are returned the front-end will need to lookup the names of the drugs in each DDI. The DDIs return URIs for the participating drugs. The map format makes this easy.

File: asclepius/README.md (hunk 3 above)
Comment (c-brenn) [reply asclepius/README.md:88]: Perhaps not needed - I kind of like the context that the outer object gives though.

**Comment by 22a** [comment]
Working on the live updates(drugs show up before ddis are queried), pls don't merge yet 💇‍♂️

File: asclepius/asclepius/main.py [hunk 9]
```diff
@@ -63,15 +63,31 @@ def ddis():
...
     return jsonify(dinto_res)
 
 
+@app.route('/uris', methods=["POST"])
+def uris():
+    """
+    """
+    params = request.get_json()
+    if params is None or 'labels' not in params or not params['labels']:
+        raise InvalidUsage("Expecting {'labels': [...]} with at least two labels")
```
Comment (c-brenn) [inline asclepius/asclepius/main.py]: this message doesn't really match up

File: panacea/web/static/js/app.js [hunk 10]
```diff
@@ -27,35 +27,49 @@ async function submitFile() {
...
         });
 
-        if (ddisResponse.ok) {
-          const data = await ddisResponse.json();
-          const ddis = data.ddis;
-          displayDdis(ddis);
+        if (urisResponse.ok) {
+          const data = await urisResponse.json();
+          const uris = Object.keys(data.uris.found);
+          const unidentifiedDrugs = Object.values(data.uris.not_found);
```
Comment (c-brenn) [inline panacea/web/static/js/app.js]: not sure if you need Object.values here.  I think not_found is just a list.

Comment (22a) [reply panacea/web/static/js/app.js]: oops lol

**Comment by 22a** [comment]
few things going on in that last commit:
* added unidentified drugs panel
* added sample pml file with silly "drugs" for manual visual testing of panels
* removed superfluous class removal in the display functions (the classes get reset at the start of the request chain)

the unidentified panel looks like this:
screen shot 2017-03-08 at 21 37 27


it also raises a few questions:
* what do we want to display to the user when the request chain is broken at the different points (when found URI's is less than `2` and when identifiable drugs is less than `0`{thinking about this, this should also probably be `2`, as there's no point getting the uri of one drug alone}) 
* do we want to make the error panel more expressive (where we can change the panel header as well)

**Comment by 22a** [comment]
also, the request chain is getting quite hairy now(v nested), perhaps we could consider breaking the chain out to separate handlers? **bleugh**

**Comment by c-brenn** [comment]
Nice work.

👍 yeah I'm all for refactoring the request chain.

I think we need a bigger discussion about the UI that's beyond the scope of this PR.

File: panacea/web/static/js/app.js [hunk 11]
```diff
@@ -27,35 +27,49 @@ async function submitFile() {
...
+          }
         } else {
-        displayError(await ddisResponse.json());
+          displayError(await urisResponse.json());
         }
-      } else {
-        displayError(await urisResponse.json());
+      }
+      else {
+        displayError({"message": "No drugs found"});
```
Comment (22a) [inline panacea/web/static/js/app.js]: this is old syntax, `{message: "bleugh"}` would be nicer, i see we have used it above

**Comment by c-brenn** [comment]
think this is ready to go - any further changes needed?

**Comment by 22a** [comment]
Looks good to 🚢  the 🐶 to me. @houli will need to approve his requested changes as I feel dismissing his review would be a breach of interpersonal trust and set a worrying precedent

**Comment by c-brenn** [comment]
Actually, the frontend changes should go in he change log.

* the unidentified drugs
* change in spinner behaviour

**Comment by 22a** [comment]
Good spot, will add those now

**Review by houli — APPROVED** [review]

**Pull request merged**